shared/api/image_decoder_win32.hpp

```diff
@@ -0,0 +1,156 @@
+                                               &pIDecoder                       // Pointer to the decoder
+    );
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed to create pIDecoder."};
+    }
+
+    // Retrieve the first bitmap frame.
+    hr = pIDecoder->GetFrame(0, &pIDecoderFrame);
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed when pIDecoder->GetFrame."};
+    }
+
+    // Now get a POINTER to an instance of the Pixel Format
+    hr = pIDecoderFrame->GetPixelFormat(&pixelFormat);
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed when pIDecoderFrame->GetPixelFormat."};
+    }
+
+    hr = _pIWICFactory->CreateComponentInfo(pixelFormat, &pIComponentInfo);
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed when pIWICFactory->CreateComponentInfo."};
+    }
+
+    // Get IWICPixelFormatInfo from IWICComponentInfo
+    IWICPixelFormatInfo2* pIPixelFormatInfo = NULL;
+    hr = pIComponentInfo->QueryInterface(__uuidof(IWICPixelFormatInfo2), reinterpret_cast<void**>(&pIPixelFormatInfo));
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed to query IWICPixelFormatInfo."};
+    }
+
+    UINT uiWidth = 0;
+    UINT uiHeight = 0;
+
+    hr = pIDecoderFrame->GetSize(&uiWidth, &uiHeight);
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: pIDecoderFrame->GetSize."};
+    }
+
+    const int height = static_cast<int>(uiHeight);
+    const int width = static_cast<int>(uiWidth);
+    const int channels = 3;  // Asks for RGB
+
+    std::vector<int64_t> output_dimensions{height, width, channels};
+    uint8_t* decoded_image_data = output.Allocate(output_dimensions);
+    if (decoded_image_data == nullptr) {
+      return {kOrtxErrorInvalidArgument, "[ImageDecoder]: Failed to allocate memory for decoded image data."};
+    }
+
+    IWICBitmapSource* pSource = pIDecoderFrame;
+
+    // Convert to 24 bytes per pixel RGB format if needed
+    if (pixelFormat != GUID_WICPixelFormat24bppRGB) {
+      IWICBitmapSource* pConverted = NULL;
+      hr = WICConvertBitmapSource(GUID_WICPixelFormat24bppRGB, pSource, &pConverted);
+      if (FAILED(hr)) {
+        return {kOrtxErrorInternal, "[ImageDecoder]: Failed when WICConvertBitmapSource."};
+      }
+
+      pSource->Release();
+      pSource = pConverted;
+    }
+
+    const int rowStride = uiWidth * sizeof(uint8_t) * channels;
+    hr = pSource->CopyPixels(NULL, rowStride, output.SizeInBytes(), decoded_image_data);
+
+    if (FAILED(hr)) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed when pConvertedFrame->CopyPixels."};
+    }
+
+    pIComponentInfo->Release();
+    pSource->Release();
+    pIWICStream->Release();
+    pIDecoder->Release();
+
+    return status;
+  }
+
+  ~DecodeImage() {
+    _pIWICFactory->Release();
+    CoUninitialize();
+  }
+
+  private:
+    IWICImagingFactory* _pIWICFactory{NULL};
```

Comment:
coding converntion: the underscore is at the end of the member name.

Reply:
sure